File: tobas_core/tobas_rqt_py/tobas_rqt_py/widgets/progress_dialog.py
from typing import override, Optional
from PyQt5.QtCore import Qt
from PyQt5.QtWidgets import QWidget, QProgressDialog

from tobas_std_tools_py.math import remap

from ..utils import qsleep


class ProgressDialog(QProgressDialog):
    """
    ===== QProgressDialogとの違い =====
    - ユーザーが他のUI要素と対話できないようにする
    - タイトルを設定
    - デフォルトで最小値に設定
    - 各操作後に画面更新のためスリープ
    - 追加メソッド
    """

    REFLESH_SLEEP = 50  # [ms]

    @override
    def __init__(self, parent: Optional[QWidget] = None, title: str = "", num_steps: int = 1) -> None:
        assert num_steps > 0

        super().__init__(parent=parent)

        self.setWindowModality(Qt.WindowModality.WindowModal)  # ユーザーが他のUI要素と対話できないようにする
        self.setWindowTitle(title)

        self._num_steps = num_steps
        self._step = 0
        self.set_step(self._step)

    @override
    def show(self, reflesh: bool = True) -> None:
        super().show()
        if reflesh:
            self.reflesh()

    @override
    def setValue(self, value: int, reflesh: bool = True) -> None:
        super().setValue(value)
        if reflesh:
            self.reflesh()

    @override
    def setLabelText(self, text: str, reflesh: bool = True) -> None:
        super().setLabelText(text)
        if reflesh:
            self.reflesh()

    def set_step(self, step: int, reflesh: bool = True) -> None:
        assert 0 <= step <= self._num_steps

        self._step = step
        value = int(remap(step, 0, self._num_steps, self.minimum(), self.maximum()))
        self.setValue(value, reflesh=reflesh)

    def progress_step(self, reflesh: bool = True) -> None:
        self._step += 1
        self.set_step(self._step, reflesh=reflesh)

    def reflesh(self) -> None:
        # 画面更新のためにメインループを進める必要がある
        # https://stackoverflow.com/questions/47879413/pyqt-qprogressdialog-displays-as-an-empty-white-window
        qsleep(self.REFLESH_SLEEP)

File: tobas_gui/tobas_qt_tools/src/widgets/progress_dialog.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 Tobas, Inc.

#include "tobas_qt_tools/widgets/progress_dialog.hpp"

#include <QApplication>
#include <QThread>

#include <tobas_math/core.hpp>

#include "tobas_qt_tools/event.hpp"

using namespace std::chrono_literals;

namespace tobas
{
namespace qt
{
ProgressDialog::ProgressDialog(const QString& title, int num_steps, QWidget* parent)
  : super(parent), num_steps_(num_steps), timer_(this)
{
  assert(num_steps > 0);

  setWindowModality(Qt::WindowModal);  // ユーザーが他のUI要素と対話できないようにする
  setWindowTitle(title);
  setStep(step_);

  connect(&timer_, &QTimer::timeout, this, &self::onTimerTimeout);
}

void ProgressDialog::show()
{
  timer_.start(250ms);
  super::show();
}

void ProgressDialog::hide()
{
  timer_.stop();
  super::hide();
}

void ProgressDialog::setLabelText(const QString& text)
{
  text_ = text;
}

void ProgressDialog::setStep(int step)
{
  step_ = std::clamp(step, 0, num_steps_);
  const auto value = math::remap(step_, 0, num_steps_, minimum(), maximum());
  setValue(value);
}

void ProgressDialog::progressStep()
{
  setStep(step_ + 1);
}

void ProgressDialog::onTimerTimeout()
{
  if (text_.isEmpty()) {
    return;
  }

  // スピナーの文字を決定
  spinner_step_ = (spinner_step_ + 1) % kSpinnerFrameSize;
  const auto spinner = kSpinnerFrames[spinner_step_];

  // スピナー部分だけ等幅で表示
  super::setLabelText(
    QString(R"(<div style="text-align:center;">%1 <span style="font-family:monospace;">%2</span></div>)")
      .arg(text_)
      .arg(spinner));
}
}  // namespace qt
}  // namespace tobas
